Finish AArch64 port

jpf91 opened the pull request:
except for TypeInfo based vaarg and some quadruple real stuff in std.math. However, everything else should be working and this should be sufficient to build a DDMD compiler.

Some commits need to go to upstream or need to be squashed, but this is nevertheless ready for review.

jpf91 committed:
Backport: AArch64 Fiber implementation

jpf91 committed:
Backport: AArch64 fix for typo in std.math

jpf91 committed:
Backport: Support for quadruple reals in core.internal.convert

jpf91 committed:
Backport: std.conv.parse changes for quadruple reals

https://github.com/dlang/phobos/commit/4c77ff8632bc086fff7cf6281c4fbf0d80037916

jpf91 committed:
Testsuite fixes for AArch64

jpf91 committed:
Backport: AArch64: Fix backtraces within Fibers

https://github.com/dlang/druntime/pull/2308
40c1108b9ad0fc573130c3ce5ab630e2e75ea047

jpf91 committed:
Backport: Quadruple real support for std.math

jpf91 committed:
Backport: Quadruple real change for std.random

jpf91 committed:
Backport: Allow small differences in real number in test

https://github.com/dlang/phobos/pull/6719
5870efca77b57d79043b3520862cbff732d433ef

dnadlinger commented on libphobos/src/std/math.d:
```diff
@@ -887,7 +887,7 @@ Lret: {}
             ];
             static immutable real[7] Q = [
-                8.650244186622719093893836740197250197602E10L
+                8.650244186622719093893836740197250197602E10L,
```
Oops – good catch!

jpf91 replied:
That was actually caught by a unittest, I wouldn't have noticed this otherwise ;-)

ibuclaw commented:
Cool, can this also be ported back to stable as well?

Would be nice to have this available so that we can test aarch64 with enable-bootstrap.

ibuclaw commented on gcc/testsuite/gdc.test/runnable/test36.d:
```diff
@@ -59,7 +63,7 @@ extern (Windows):
 {
         auto inst= new invarianttest;
-        int esp;
+        size_t esp;
```
I think that this breaks x86

jpf91 replied:
Hmm, OK. Is it actually intended that this test saves only 32 bit of the 64bit sp on 64 bit architectures? I guess I'll just change that back to `int` and change the AArch64 asm to save only 32 bit, as is done for X86?

ibuclaw replied:
```
test36.d: Assembler messages:
test36.d:39: Error: unsupported instruction `mov'
test36.d:71: Error: unsupported instruction `mov'
test36.d:94: Error: unsupported instruction `mov'
```
Just use `%%rsp` for the 64bit path.

ibuclaw replied:
There's also X32 to consider, which will complain as well, but that's been untested for years, and probably has more problems to fix.

jpf91 replied:
I just save 32 bits for AArch64 now (like X86_32), leaving other targets unaffected.

jpf91 commented:
@ibuclaw This already targets stable, it'll need to be ported to master ;-)

ibuclaw commented:
*blinks*

jpf91 committed:
Backport: AArch64: Fix typo in std.math

https://github.com/dlang/phobos/pull/6719
00301f0f0b2d5ca7c5c3bf77a9902285d1c2f091

ibuclaw commented on libphobos/libdruntime/core/internal/convert.d:
```diff
@@ -35,13 +35,30 @@ const(ubyte)[] toUbyte(T)(ref T val) if(is(Unqual!T == float) || is(Unqual!T ==
         size_t off_bits  = 0;
 
-        for(; off_bytes < FloatTraits!T.MANTISSA/8; ++off_bytes)
+        // Quadruples won't fit in one ulong, so check for that.
```
You probably want to take the implementation from master, rather than my version from the riscv patches.

jpf91 replied:
OK, uses the master implementation now.

ibuclaw commented on gcc/testsuite/gdc.test/runnable/test42.d:
```diff
@@ -4247,6 +4247,8 @@ int bug3809()
         else version(ARM)
             asm { "nop"; }
+        else version(AArch64)
+            asm { "nop"; }
```
I wonder, can we just have `asm { ""; }` here?

ibuclaw commented on gcc/testsuite/gdc.test/runnable/testsafe.d:
```diff
@@ -215,6 +215,8 @@ void inlineasm()
         else version(ARM)
             static assert(!__traits(compiles, { asm { "nop"; } }() ));
+        else version(AArch64)
+            static assert(!__traits(compiles, { asm { "nop"; } }() ));
```
Likewise here, `asm { ""; }` should be rejected as well, and is not platform dependent.

jpf91 replied:
done.

jpf91 commented:
Rebased, backported the changes which were now committed upstream and addressed the "asm { ""; }" comments. I'll run the testsuite manually once again, then try marking tests as `{ dg-do run { xfail aarch64*-*-* } }` and then we should be able to enabled aarch64 testing in buildkite.

jpf91 committed:
Add support to skip tests in libphobos unittests (mark UNSUPPORTED)

jpf91 committed:
Mark unsupported AArch64 tests

jpf91 commented:
@ibuclaw This should be ready.

The most interesting commit is probably https://github.com/D-Programming-GDC/GDC/pull/744/commits/6d9f93c89387a6773614f533491bfebafa1b2b5c : It extends our libphobos.unittest test suite to allow skipping tests / marking them as UNSUPPORTED using GCC selectors: https://gcc.gnu.org/onlinedocs/gcc-8.1.0/gccint/Selectors.html#Selectors

With this, it's now possible to mark all failing math unittests and testsuite tests `(dg-skip-if)` as in https://github.com/D-Programming-GDC/GDC/pull/744/commits/50066191d49ccc11bf1fe46878fe18b9cb391b08 . This means we should now be able to enable CI for AArch64 and get a pass result, while still properly showing unsupported tests. This should make it much easier to catch regressions for partially supported targets.

So could you enable AArch64 testsuite / unittests on Buildkite to test this?

ibuclaw reviewed (approved):
OK

ibuclaw commented:
Do you want to post this to gcc-patches, or shall I?

jpf91 commented:
> Do you want to post this to gcc-patches, or shall I?

You're already familiar with the process, so I'll leave that up to you ;-)

jpf91 commented:
Can we merge this PR or should this wait until gcc patches have been merged / reviewed?

ibuclaw commented:
You can merge. I need to find some time to synchronize this with gcc trunk though, as they have started to diverge.

jpf91 commented:
OK, thanks!
